### ISSUE TITLE
gatherResults should have an optional consumeErrors

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5159]
|[<img alt="dustin's avatar" src="https://avatars.githubusercontent.com/u/28673?s=50" width="50" height="50">](https://github.com/dustin)| @dustin reported|
|-|-|
|Trac ID|trac#5159|
|Type|enhancement|
|Created|2011-06-26 02:11:23Z|
|Branch|https://github.com/twisted/twisted/tree/gatherresults-consumeerrors-5159|

It's been suggested a few times that those who need to use gatherResults, but want any failures in the underlying Deferreds to be handled exclusively by the DeferredList should write their own gatherResults method.

What if this method had an optional consumeErrors=False keyword arugment?

Attachments:

* [bug5159-r1.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a81/a814197c359e9b098b07c33abcde7728e9b4f4c3/019fffee2905db6714ecf349846588ee7e21503c.patch) (1978 bytes) - added by dustin on 2011-06-26 02:20:01Z - bug5159-r1.patch
* [bug5159-r3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a81/a814197c359e9b098b07c33abcde7728e9b4f4c3/8ee7196e29d692e773adc0e322c092960bab4b91.patch) (6276 bytes) - added by dustin on 2011-07-06 01:10:41Z - bug5159-r3.patch

<details><summary>Searchable metadata</summary>

```
trac-id__5159 5159
type__enhancement enhancement
reporter__dustin dustin
priority__normal normal
milestone__ 
branch__branches_gatherresults_consumeerrors_5159 branches/gatherresults-consumeerrors-5159
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1309054283000000 1309054283000000
changetime__1312229198000000 1312229198000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="dustin's avatar" src="https://avatars.githubusercontent.com/u/28673?s=50" width="50" height="50">](https://github.com/dustin)<a name="note_2"></a>|@dustin commented|
|-|-|

Links for the "few times" in the description:
 * http://www.twistedmatrix.com/pipermail/twisted-python/2008-November/018639.html
 * [#4627](https://github.com/twisted/twisted/issues/4627)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

This seems like a duplicate of [#5061](https://github.com/twisted/twisted/issues/5061).

**Comment by twisted-trac** [comment]
|[<img alt="dustin's avatar" src="https://avatars.githubusercontent.com/u/28673?s=50" width="50" height="50">](https://github.com/dustin)<a name="note_4"></a>|@dustin commented|
|-|-|

Not quite (this is a simpler solution to a simpler problem), but I'm sorry I didn't find that bug before filing this one.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @dustin|
|-|-|

Okay.  Thanks for the quick follow-up (sorry about mine being so delayed).

  1. The `gatherResults` docstring change for `consumeErrors` seems a little round-about to me.  Do you think it would be better if it just described the behavior of `consumeErrors`?  The current form of trying to proscribe usage seems too narrow to me; it might serve as part of a good example, but not as the canonical documentation for the parameter (I might even say `DeferredList`'s docstring should be canonical, and `gatherResults` could just refer to that, but I also want to get rid of `DeferredList` ;).
  1. The test method name should follow the `test_something` convention, rather than the older `testSomething` convention.
  1. The test method should have a docstring.
  1. The test method would probably be better if it tested the error handling behavior rather than just checking the flag.  Maybe we'll be able to get away with re-implementing `gatherResults` in terms of the replacement for `DeferredList` someday, and this attribute won't even exist anymore; the test could be indifferent to that change if it only checks behavior.
  1. It would also be very nice if we had some howto-style documentation that covered `gatherResults` and this new flag for it.  Do you feel like putting together a very quick addition for one of the Deferred howtos to demonstrate this API and the difference between `consumeErrors=False` and `consumeErrors=True`?  Fortunately `doc/core/howto/defer.xhtml` does talk about `DeferredList` and its usage of `consumeErrors` already, so an extension to that section might not be too hard.

Thanks again!  I don't think there should be any problem merging a patch that addresses these issues.

**Comment by twisted-trac** [comment]
|[<img alt="dustin's avatar" src="https://avatars.githubusercontent.com/u/28673?s=50" width="50" height="50">](https://github.com/dustin)<a name="note_6"></a>|@dustin removed owner|
|-|-|

1: I'm not entirely sure what you mean by "trying to proscribe usage" - does the attached address your concerns?

2-4: done

5: I added a new subsection under DeferredList, treating gatherResults as a shortcut.  The docs can change later if gatherResults grows into something that stands alone.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [32389]) Branching to 'gatherresults-consumeerrors-5159'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [32390]) Apply dustin's patch

refs [#5159](https://github.com/twisted/twisted/issues/5159)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [32392]) Address simple coding standard points

  1. Docstring formatting
  2. More description of the behavior being tested in the docstring
  3. camelCase instead of under_scores in variable names
  4. assertIsInstance instead of failUnless(check(...))

refs [#5159](https://github.com/twisted/twisted/issues/5159)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set owner to @exarkun|
|-|-|

Thanks.  I made some changes, some linked above, others:

  1. Changed the docs to say >= 11.1 instead of > 11.0
  1. Referred back to `DeferredList` in the new `gatherResults` prose docs for the meaning of the `consumeErrors` parameter.
  1. Made the new test not return a Deferred, since it completes synchronously
  1. Changed the wording in the news fragment to follow the [ReviewProcess#Newsfiles style guide].

The [build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/gatherresults-consumeerrors-5159) look good so far.  Assuming the last one completes successfully, I'll merge.  Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set status to `closed`|
|-|-|

(In [32405]) Merge gatherresults-consumeerrors-5159 

Author: dustin
Reviewer: exarkun
Fixes: [#5159](https://github.com/twisted/twisted/issues/5159)

Add a `consumeErrors` parameter to `twisted.internet.defer.gatherResults`
with the same meaning as the parameter of the same name accepted by
`DeferredList`.
